### PR TITLE
[WPF][XamMac][GtkMac] A11y helper

### PR DESCRIPTION
**Pull request by daserge**
* Allows to check whether accessibility means are in use (Narrator/3rd party tools on Windows, VoiceOver on macOS)
* Adds `Accessible.MakeAnnouncement (string message, bool polite = false)` method for making a custom announcement for user per request (uses LiveRegion on Windows and `NSAccessibility.PostNotification` on macOS)

WPF version requires target framework version .NET 4.7.2

### CODE REVIEW

File: Xwt.XamMac/Xwt.Mac/AccessibilityHelper.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+			var block = new BlockLiteral ();
+
+			SwizzledAccessibilitySetValueForAttributeDelegate d = accessibilitySetValueForAttribute;
+			block.SetupBlock (d, null);
+			var imp = imp_implementationWithBlock (ref block);
+			method_setImplementation (accessibilitySetValueForAttributeMethod, imp);
+
+			accessibilityInUse = CFPreferences.GetAppBooleanValue ("voiceOverOnOffKey", "com.apple.universalaccess");
+			Console.WriteLine ("Xwt AccessibilityHelper initial accessibilityInUse: " + accessibilityInUse);
```
Comment (garuma) [inline Xwt.XamMac/Xwt.Mac/AccessibilityHelper.cs]: Leftover C.W.L

File: Xwt.XamMac/Xwt.Mac/AccessibilityHelper.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		static void accessibilitySetValueForAttribute (IntPtr block, IntPtr self, IntPtr valueHandle, IntPtr attributeHandle)
+		{
+			var d = Marshal.GetDelegateForFunctionPointer (originalAccessibilitySetValueForAttributeMethod, typeof (AccessibilitySetValueForAttributeDelegate));
+			d.DynamicInvoke (self, Selector.GetHandle ("accessibilitySetValue:forAttribute:"), valueHandle, attributeHandle);
+
+			var val = (NSNumber)ObjCRuntime.Runtime.GetNSObject (valueHandle);
+
+			bool previousValue = accessibilityInUse;
+			accessibilityInUse = val.BoolValue;
+			Console.WriteLine ("Xwt AccessibilityHelper accessibilityInUse: " + accessibilityInUse);
```
Comment (garuma) [inline Xwt.XamMac/Xwt.Mac/AccessibilityHelper.cs]: ditto

File: Xwt.XamMac/Xwt.Mac/AccessibilityHelper.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			get
+			{
+				if (!initialized)
+					SwizzleNSApplicationAccessibilitySetter ();
+
+				return accessibilityInUse;
+			}
+		}
+
+		public static void MakeAnnoucement (string message)
```
Comment (Therzok) [inline Xwt.XamMac/Xwt.Mac/AccessibilityHelper.cs]:
```suggestion
		public static void MakeAnnouncement (string message)
```

File: Xwt.XamMac/Xwt.Mac/AccessibleBackend.cs [hunk 4]
```diff
@@ -284,5 +284,10 @@ public void DisableEvent (object eventId)
 				}
 			}
 		}
+
+		public void MakeAnnouncement (string message)
+		{
+			AccessibilityHelper.MakeAnnoucement (message);
```
Comment (Therzok) [inline Xwt.XamMac/Xwt.Mac/AccessibleBackend.cs]:
```suggestion
			AccessibilityHelper.MakeAnnouncement (message);
```

File: Xwt.Gtk.Mac/GtkMacAccessibleBackend.cs [hunk 5]
```diff
@@ -147,5 +148,10 @@ public override void RemoveAllChildren ()
 
 			nsa.AccessibilityChildren = null;
 		}
+
+		public override void MakeAnnouncement (string message)
+		{
+			AccessibilityHelper.MakeAnnoucement (message);
```
Comment (Therzok) [inline Xwt.Gtk.Mac/GtkMacAccessibleBackend.cs]:
```suggestion
			AccessibilityHelper.MakeAnnouncement (message);
```

**Comment by sevoku** [comment]
Xwt.Wpf has still the bumped framework. My feeling is that it's a bit too complex, can't we use separate configurations with and without the framework bump and conditionally enable/disable the feature that depends on 4.7? We'd need to update appveyor and nuspecs to build and include both in the packages, but this might be a more robust solution than injecting an additional project.

cc @slluis

**Comment by daserge** [comment]
>Xwt.Wpf has still the bumped framework.

Good catch, thanks! This is now reverted.

>use separate configurations with and without the framework bump and conditionally enable/disable the feature that depends on 4.7

Do you know how this could be implemented?
Or could you please give a reference?

**Comment by sevoku** [comment]
@daserge with https://github.com/mono/xwt/pull/954 we can now conditionally include the announcement code in Xwt.Wpf directly using `#ifdef NET47`: https://github.com/mono/xwt/blob/master/Xwt.WPF/Xwt.WPF.csproj, so we don't need the helper assemblies anymore.

File: Xwt.Gtk.Mac/Xwt.Gtk.Mac.csproj [hunk 6]
```diff
@@ -78,5 +79,9 @@
       <Project>{92494904-35FA-4DC9-BDE9-3A3E87AC49D3}</Project>
       <Name>Xwt</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Xwt.XamMac\Xwt.XamMac.csproj">
```
Comment (sevoku) [inline Xwt.Gtk.Mac/Xwt.Gtk.Mac.csproj]: This is not right, `Xwt.Gtk.Mac` should not depend on `Xwt.XamMac`, also problematic with packages.

File: Xwt.Gtk.Mac/GtkMacAccessibleBackend.AccessibilityHelper.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+using Xwt.GtkBackend;
+
+namespace Xwt.Gtk.Mac
+{
+	public partial class GtkMacAccessibleBackend : AccessibleBackend
+	{
+		static Xwt.Mac.AccessibleBackend xamMacA11yHelper;
+		static GtkMacAccessibleBackend ()
+		{
+			xamMacA11yHelper = new Xwt.Mac.AccessibleBackend ();
```
Comment (sevoku) [inline Xwt.Gtk.Mac/GtkMacAccessibleBackend.AccessibilityHelper.cs]: I'd rather move the static part of Xwt.XamMac/Xwt.Mac/AccessibleBackend.AccessibilityHelper.cs to a separate file, linked in this project and then copy the instance code here.

File: Xwt.Gtk.Mac/MacPlatformBackend.cs [hunk 8]
```diff
@@ -27,22 +27,21 @@
...
 using Xwt.Backends;
 using AppKit;
-using Xwt.Mac;
 
 namespace Xwt.Gtk.Mac
 {
 	public class MacPlatformBackend: GtkPlatformBackend
 	{
-		public override void Initialize (ToolkitEngineBackend toolit)
+		public override void Initialize (ToolkitEngineBackend toolkit)
```
Comment (sevoku) [inline Xwt.Gtk.Mac/MacPlatformBackend.cs]: this change is unrelated, either revert it or create a separate commit.

File: Xwt.WPF/Xwt.WPFBackend/AccessibleBackend.cs [hunk 9]
```diff
@@ -1,18 +1,17 @@
 using System;
 using System.Collections.Generic;
-using System.Linq;
-using System.Text;
+using System.Threading.Tasks;
```
Comment (sevoku) [inline Xwt.WPF/Xwt.WPFBackend/AccessibleBackend.cs]: Usings are unrelated here, we can clean them up, but adding new ones doesn't make sense.

File: Xwt.WPF/Xwt.WPFBackend/WPFEngine.cs [hunk 10]
```diff
@@ -37,7 +37,6 @@
 using Xwt.Backends;
 using Xwt.Drawing;
 
-
```
Comment (sevoku) [inline Xwt.WPF/Xwt.WPFBackend/WPFEngine.cs]: Unrelated change

File: Xwt.sln [hunk 11]
```diff
@@ -295,7 +476,4 @@ Global
 		$0.CSharpFormattingPolicy = $5
 		$5.scope = text/x-csharp
 	EndGlobalSection
-	GlobalSection(SolutionProperties) = preSolution
-		HideSolutionNode = FALSE
-	EndGlobalSection
 EndGlobal
```
Comment (sevoku) [inline Xwt.sln]: Is this change required?

File: Xwt/Xwt.Backends/ToolkitEngineBackend.cs [hunk 12]
```diff
@@ -31,6 +31,7 @@
 using Xwt.Drawing;
 
 using System.Collections.Generic;
+using Xwt.Accessibility;
```
Comment (sevoku) [inline Xwt/Xwt.Backends/ToolkitEngineBackend.cs]: Unrelated?

File: Xwt/Xwt/Application.cs [hunk 13]
```diff
@@ -107,7 +107,7 @@ public static void Initialize (string backendType)
 			toolkit.SetActive ();
 			engine = toolkit.Backend;
 			mainLoop = new UILoop (toolkit);
-
+
```
Comment (sevoku) [inline Xwt/Xwt/Application.cs]: Unrelated?

**Comment by daserge** [comment]
@sevoku addressed, will now test it with clients

File: TestApps/GtkOnMacTest/Program.cs [hunk 14]
```diff
@@ -34,6 +34,7 @@ class MainClass
 	{
 		public static void Main (string[] args)
 		{
+			ObjCRuntime.Dlfcn.dlopen("/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/libxammac.dylib", 0);
```
Comment (Therzok) [inline TestApps/GtkOnMacTest/Program.cs]: Why is this needed?

Comment (daserge) [reply TestApps/GtkOnMacTest/Program.cs]: Otherwise it fails to find the libxammac.dylib

Comment (Therzok) [reply TestApps/GtkOnMacTest/Program.cs]: I think that's because we only copy Xamarin.Mac.dll:
https://github.com/mono/xwt/blob/275741a3ad577329fb8979285c4f8d2ef0112123/TestApps/GtkOnMacTest/GtkOnMacTest.csproj#L35

Comment (sevoku) [reply TestApps/GtkOnMacTest/Program.cs]: see my change here: https://github.com/mono/xwt/commit/415e9e57dbd9f58c8607e628d788a9fc90ab7ad9
Afair that's required because of the project type.

Comment (daserge) [reply TestApps/GtkOnMacTest/Program.cs]: I've removed those changes - anyways they are unrelated to the PR, so you can apply them locally if you want to test
